python/tk_multi_publish2/api/manager.py

```diff
@@ -0,0 +1,627 @@
+        publish tree.
+
+        By default, the method will iterate over the manager's publish tree,
+        finalizing all active tasks on all active items. To process tasks in a
+        different way (different order or different criteria) you can provide
+        a custom ``task_generator`` that yields :class:`~PublishTask` instances.
+
+        For example, to finalize all items in the tree that have a
+        ``generate_report`` flag set in their properties dictionary, you could
+        do the following:
+
+        .. code-block:: python
+
+            def report_tasks_generator(publish_tree):
+
+                for item in publish_tree:
+                    if item.properties.get("generate_report"):
+                        for task in item.tasks:
+                            yield task
+
+            publish_manager.validate(task_generator=report_tasks_generator)
+
+        :param task_generator: A generator of :class:`~PublishTask` instances.
+        """
+
+        # calling code can supply its own generator for tasks to process. if not
+        # supplied, we'll use our own generator.
+        if not task_generator:
+            task_generator = self._task_generator()
+
+        task = task_generator.next()
+
+        # now begin iterating over tasks supplied by the generator
+        while task:
+
+            finalize_exception = None
+
+            # do the actual finalize
+            # the UI's generator to update the display of the task as it is
+            # being processed.
+            try:
+                task.finalize()
+            except Exception, e:
+                finalize_exception = e
+
+            # send the the exception (or None) back to the generator and get the
+            # next task. this is a bit annoying since send() returns the next
+            # value of the generator. which is why we're using `while` instead
+            # of a for loop with the generator.
+            try:
+                task = task_generator.send(finalize_exception)
+            except StopIteration:
+                break
+
+        # execute the post finalize method of the phase phase hook
+        self._post_phase_hook.post_finalize(self.tree)
+
+    @property
+    def context(self):
+        """Returns the execution context of the manager."""
+        return self._bundle.context
+
+    @property
+    def logger(self):
+        """Returns the logger used during publish execution."""
+        return self._logger
+
+    @property
+    def collected_files(self):
+        """
+        Returns a list of file paths for all items collected via the
+        collect_files method.
+        """
+        collected_paths = []
+        for item in self.tree.persistent_items:
+            if self.PROPERTY_KEY_COLLECTED_FILE_PATH in item.properties:
+                collected_paths.append(
+                    item.properties[self.PROPERTY_KEY_COLLECTED_FILE_PATH])
+
+        return collected_paths
+
+    @property
+    def tree(self):
```

Comment:
Made the `tree` structure a public interface to allow us to pass it to hooks like the new `post_phase` hook.